### PR TITLE
Nft market implementation

### PR DESCRIPTION
**Pull request by jununifi**

### CODE REVIEW

**Comment by taiki1frsy** [comment]
Just for the information, when you support sdk .46-rc1, you need to modify two function names in app/test_helper.go, which are:
sdk.AccAddressFromHex -> sdk.AccAddressFromHexUnsafe
helper.GenTx -> helper.GenSignedMockTx

https://github.com/UnUniFi/chain/blob/f742d74cb0459ea2cbbdd6f586d7d9435805b7dc/app/test_helper.go#L270
https://github.com/UnUniFi/chain/blob/f742d74cb0459ea2cbbdd6f586d7d9435805b7dc/app/test_helper.go#L373

File: proto/gen.sh [hunk 1]
```diff
@@ -35,13 +35,13 @@ Mgoogle/protobuf/any.proto=github.com/cosmos/cosmos-sdk/codec/types:. \
...
-  --doc_opt=./docs/protodoc-markdown.tmpl,proto-docs.md \
-  $(find "$(pwd)/proto" -maxdepth 5 -name '*.proto')
-go mod tidy
+# buf protoc \
+#   -I "proto" \
+#   -I "proto-thirdparty" \
+#   --doc_out=./docs/core \
+#   --doc_opt=./docs/protodoc-markdown.tmpl,proto-docs.md \
+#   $(find "$(pwd)/proto" -maxdepth 5 -name '*.proto')
+# go mod tidy
```
[the remark this answers is not in the document]
Comment (jununifi) [reply proto/gen.sh:44]: The changes will be reverted

File: x/nftmarket/keeper/bid.go [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+	store.Delete(types.NftBidKey(bid.IdBytes(), bidder))
+	store.Delete(types.AddressBidKey(bid.IdBytes(), bidder))
+}
+
+func getCancelledBidTimeKey(timestamp time.Time) []byte {
+	timeBz := sdk.FormatTimeBytes(timestamp)
+	timeBzL := len(timeBz)
+	prefixL := len(types.KeyPrefixNftBidCancelled)
+
+	bz := make([]byte, prefixL+8+timeBzL)
```
[the remark this answers is not in the document]
Comment (jununifi) [reply x/nftmarket/keeper/bid.go:116]: It's a common strategy for managing time bytes

File: x/nftmarket/keeper/bid.go [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+	bids := []types.NftBid{}
+	for ; it.Valid(); it.Next() {
+		bid := types.NftBid{}
+		k.cdc.MustUnmarshal(it.Value(), &bid)
+		bids = append(bids, bid)
+	}
+	return bids
+}
+
+func (k Keeper) GetMaturedCancelledBids(ctx sdk.Context, endTime time.Time) []types.NftBid {
```
[the remark this answers is not in the document]
Comment (jununifi) [reply x/nftmarket/keeper/bid.go:149]: After cancelled, it's not returned instantly but it's returned after some time,  matured means finished waiting, once bid cancelled, cancelled bid queue is created and after some time, it is put on matured cancelled bids.

[the remark this answers is not in the document]
Comment (mkXultra) [reply x/nftmarket/keeper/bid.go:149]: got it!

**Pull request merged**